File: INFORMÁTICA/PROGRAMACIÓN Y DISEÑO ORIENTADO A OBJETOS/PRÁCTICAS/PROYECTOS/EXÁMENES/examen_2_19-20_definitivo/deepspace/src/deepspace/SpecificDamage.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package deepspace;

import java.util.ArrayList;
/**
* Hereda de la clase abstracta Damage
* Versión con tipos de armas de Damage
*/
public class SpecificDamage extends Damage {
    
    //Atributos
    //Lista con los tipos de armas que se van a perder.
    private ArrayList<WeaponType> weapons;
    
    //Constante para indicar desuso.
    private static final int NOUSE=-1;
    
    //Constructor
    SpecificDamage(ArrayList<WeaponType> wl, int s){
        super(s);
        
        if(wl!=null){
            weapons=new ArrayList<>(wl);
        }else{
            weapons=new ArrayList<>();
        }
    }
    
    @Override
    public Damage copy(){
        return (new SpecificDamage(this.weapons,this.getNShields()));
    }
    
    //Consultores
    
    public ArrayList<WeaponType> getWeapons(){return weapons;}
    
    
    //Interfaz
    @Override
    public SpecificDamageToUI getUIversion(){return (new SpecificDamageToUI(this));}
    
    private int arrayContainsType(ArrayList<Weapon> w, WeaponType t){
        
        int index=NOUSE;
        boolean seguir=true;
        for (int i=0; i<w.size() && seguir; i++){
            if (w.get(i).getType()==t){
                
                index=i;
                seguir=false;
            }
        }
        
        return index;
        
    }
    
    //Redefinimos Adjust
    @Override
    public SpecificDamage adjust(ArrayList<Weapon> w, ArrayList<ShieldBooster> s){
        
        int shields_copia;
        //Ajustamos los escudos
        if(getNShields()>s.size())
            shields_copia=s.size();
        else
            shields_copia=getNShields();  //Se usa el del padre
        
        ArrayList<WeaponType> weapons_copia=new ArrayList<>();
        ArrayList<Weapon> w_aux=new ArrayList<>(w);
        /*
        Si un tipo de arma esta en w_aux la añadimos a weapons_copia,
        si no esta no la añadimos ya que entonces perderiamos tipos de armas
        que no estan en w
        */
        for(int i=0; i<weapons.size(); i++){

            int indice=arrayContainsType(w_aux, weapons.get(i));
            //Si el tipo de weapon esta en el array w_aux la añadimos a las
            //armas que se pueden perder(las de damage)
            if(indice!=-1){  
                weapons_copia.add(weapons.get(i));
                w_aux.remove(indice);  //Lo borramos porque ya hemos encontrado este tipo de arma
            }

        }

        return (new SpecificDamage(weapons_copia, shields_copia));
        
    }
    
    //Redifinimos hasNoEffect
    @Override
    public boolean hasNoEffect(){
        
        return (getNShields()==0 && weapons.isEmpty());
    
    }
    
    @Override
    public void discardWeapon(Weapon w){
        
        int index= weapons.indexOf(w.getType());
        if (index>=0) weapons.remove(index); 
        
    }
    
    @Override
    public String toString(){
        
        String weapons_string= "WeaponsType: ";
        for(int i=0; i<weapons.size(); i++){
            weapons_string += weapons.get(i).toString() + " ";
        }
        
        return ("SpecificDamage:\n" + "nShields: "+ getNShields()+ "\n" + 
                   weapons_string + "\n");
    }
    
    
    
    
}

File: INFORMÁTICA/PROGRAMACIÓN Y DISEÑO ORIENTADO A OBJETOS/PRÁCTICAS/PROYECTOS/EXÁMENES/examen_22-23_extraordinaria/deepspace_rb/SpecificDamage.rb
#encoding: UTF-8

require_relative 'Damage'
require_relative 'SpecificDamageToUI'

module Deepspace
    
    #Versión con número de armas de 
    #Hereda de Damage
    class SpecificDamage < Damage

        @NOUSE=-1     #Atributo de instancia(porque estoy en ámbito de clase) de clase( necesitamos un consultor para acceder a el en un metodo de instancia)
                        #otra opcion es hacerlo de clase(dos @@)

        def initialize(wl, s)
            super(s)

            @weapons=wl
        end

        #Necesitamos un metodo de clase para acceder al altributo de instancia
        def self.NOUSE    
            @NOUSE
        end

        #Hacemos publico el metodo new
        public_class_method :new

        #Método de instancia copy que devolverá una copia del objeto que reciba el mensaje. 
        #Cuando sea necesario crear una copia de un objeto que represente daños, 
        #simplemente se llamará a este método copy de la instancia que se desea copiar.
        def copy 
            return (SpecificDamage.new(@weapons, nShields()))
        end

        def weapons
            @weapons
        end

        #Capa que permite conectar el modelo con la interfaz de 
        #usuario manteniendo cierto nivel de aislamiento
        def getUIversion()
            return (SpecificDamageToUI.new(self))
        end  


        #Devuelve el indice de la primera arma de w, cuyo tipo coincida con t.
        # Si no se encuentra devuelve -1
        def arrayContainsType(w,t)
            index=SpecificDamage.NOUSE   #Si pusiesemos index=@NOUSE NOUSE seria otra variable distinta, en este caso de instancia
            i=0
            seguir=true
            while i<w.size() && seguir
                if(w.at(i).type==t)
                    index=i
                    seguir=false  #Para salirnos del bucle
                end
                i += 1
                
            end        
            
               
            return index
        end

        #Devuelve una versión ajustada del objeto a w y s.
        #Se devuelve una copia del mismo pero reducida si es necesario para que no 
        #implique perder armas o potenciadores de escudos que no están en las
        #colecciones de los parámetros.

        #@param w Array de weapons.
        #@param s Array de shieldBoosters
        def adjust(w, s)
            
        
            #Ajustamos los escudos
            #if(@nShields>s.size)
            #    shields_copia=s.size()
            #else
            #    shields_copia=@nShields
            #end

            shields_copia = super(s)

            #Ajustamos las armas
            
            weapons_copia=Array.new()      #Array donde vamos a almacenar la solucion
            w_aux=Array.new(w)
            

            #Si un tipo de arma esta en w_aux la añadimos a weapons_copia,
            #si no esta no la añadimos ya que entonces perderiamos tipos de armas
            #que no estan en w
        

            @weapons.each do |element|
                indice=arrayContainsType(w_aux, element)
                #Si el tipo de weapon esta en el array w_aux la añadimos a las
                #armas que se pueden perder(las de damage)
                if(indice!=SpecificDamage.NOUSE)
                    
                    weapons_copia.push(element)
                    w_aux.delete_at(indice)  #Lo borramos porque ya hemos encontrado este tipo de arma
                end
            end
            return SpecificDamage.new(weapons_copia, shields_copia)
            
        end

        #Devuelve true si el daño no implica la pérdida de ningún tipo de accesorio 
        #(armas o potenciadores de escudo).
        def hasNoEffect()
            
            return (@weapons.size == 0 && nShields()==0)

        end

        #Disminuye, si es posible el numero de armas
        def discardWeapon(w)
        
            i= @weapons.index(w.type)     #Devuelve el primer indice que tiene el objeto(el ultimo con rindex)
            if (i!=nil) 
                #Devuelve el elemento eliminado.
                @weapons.delete_at(i)
            end
        end

        def to_s()
        
            weapons_string= "Weapons: "
            for i in 0..(@weapons.size-1) do
                weapons_string += @weapons.at(i).to_s() + "\n"
            end

            "SpecificDamage:\n" + "nShields: "+ nShields().to_s + "\n" + weapons_string + "\n"
        end





    end
end
